services/src/main/java/com/chatbot/services/AsyncService.java

```diff
@@ -0,0 +1,151 @@
+import com.google.protobuf.Value;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+
+@Service
+public class AsyncService {
+
+  @Autowired
+  private HangoutsMessageSender hangoutsMessageSender;
+  @Autowired
+  private IDMapping iDMapping;
+
+  private static final String IMAGES_RECEIVED_MESSAGE = "The images have been received!";
+  private static final String THANKS_FOR_ADDING_MESSAGE = "Thank You for Adding me";
+  private static final String NOT_EXPECTING_IMAGE_MESSAGE =
+      "Sorry, we were not expecting any attachements from you.";
+  
+  @Async("asyncExecutor")
+  void hangoutsAsyncHandler(final ChatServiceRequest chatServiceRequest) throws Exception {
+    final String spaceID = chatServiceRequest.getSender().getChatClientGeneratedId();
+    switch (chatServiceRequest.getRequestType()) {
+      case ADDED:
+        hangoutsMessageSender.sendMessage(spaceID, THANKS_FOR_ADDING_MESSAGE);
+        iDMapping.addNewMapping(chatServiceRequest.getSender().getChatClientGeneratedId(),
+            chatServiceRequest.getSender().getUserId(), chatServiceRequest.getChatClient());
+        break;
+      case REMOVED:
+        break;
+      case MESSAGE:
+        handleMessageEvent(chatServiceRequest);
+        break;
+      default:
+        break;
+    }
+  }
+
+  void handleMessageEvent(final ChatServiceRequest chatServiceRequest) throws Exception {
+    final String spaceID = chatServiceRequest.getSender().getChatClientGeneratedId();
+    // The spaceID of the user is used as the sessionID for hangouts
+    final DialogflowConversation dialogflowConversation =
+        new DialogflowConversation(System.getenv("projectID"), spaceID);
+    if (chatServiceRequest.getUserMessage().getAttachmentsCount() == 0) {
+      final Value userID = Value.newBuilder()
+          .setStringValue(chatServiceRequest.getSender().getUserId()).build();
+      final Struct payload = Struct.newBuilder()
+          .putFields("userID", userID)
+          .build();
+      final QueryResult queryResult = dialogflowConversation.sendMessage(
+          chatServiceRequest.getUserMessage().getText(), payload);
+      final String response = queryResult.getFulfillmentText();
+      if(ChatServiceConstants.LIST_OF_INTENTS_WITH_INTERACTIVE_RESPONSE
+          .contains(queryResult.getIntent().getDisplayName())) {
+        hangoutsMessageSender.sendCardMessage(spaceID, response);  
+      } else {
+        hangoutsMessageSender.sendMessage(spaceID, response);
+      }
+    } else {
+      final List<String> currentContextList = dialogflowConversation.getCurrentContexts();
+      if (currentContextList.contains(ChatServiceConstants.EXPECTING_IMAGES_CONTEXT)) {
+        // send images to backend
+        hangoutsMessageSender.sendMessage(
+          chatServiceRequest.getSender().getChatClientGeneratedId(), IMAGES_RECEIVED_MESSAGE);
```

Comment:
Add a TODO here stating, we will be sharing these images with backend.
Same for other backend interactions as well